src/framework/constraints.ml

```diff
@@ -676,8 +676,15 @@ struct
   let tf_proc var edge prev_node lv e args getl sidel getg sideg d =
     let ctx, r, spawns = common_ctx var edge prev_node d getl sidel getg sideg in
     let functions =
-      let ls = ctx.ask (Queries.EvalFunvar e) in
-      Queries.LS.fold (fun ((x,_)) xs -> x::xs) ls []
+      match e with
+      | Lval (Var v, NoOffset) ->
```

Comment:
I was a bit skeptical here, because of code such as this one:
~~~C
int f(int x) {
  return x;
}

int main() {
  int (*ptr)(int) = &f;
  ptr(3);
}
~~~

But CIL adds the implicit dereference, so it also works here.

~~~C
  (*ptr)(3);
~~~

Are there any similar cases we should have a look at?

Reply:
That's a great example I didn't think of, but I guess we're lucky CIL does that. I was a bit unsure if this would always be correct.

I think if you follow `EvalFunvar (Lval (Var v, NoOffset))` from `query` through `eval_rv` and `eval_lv`, then it ends up returning a singleton address set with `(v, NoOffset)` in it. So it should be the maximal inlining of that logic, which also depends on CIL adding the dereference (i.e. `Mem` constructor).

I guess it'd be possible to add some extra conditions here, like `v` needs to be the `svar` of some `fundec` but AFAIK base doesn't do anything like that at any step. And it would be less general because no special functions (e.g. `assert`) can be called without base activated.

Reply:
I think in CIL the explicit dereference is being added here: https://github.com/goblint/cil/blob/94ba9835b86977e73fb9b6ea0c09adc43f716a43/src/frontc/cabs2cil.ml#L4203-L4214.